[Frontend] Expand tools even if tool_choice="none"

okdshin opened the pull request:
# Add option to include tool definitions even when tool_choice is 'none'

## Summary
This PR adds a new command-line option `--expand-tools-even-if-tool-choice-none` which allows including tool definitions in prompts even when `tool_choice='none'`.

## Motivation
In the current implementation, when `tool_choice` is set to `'none'`, all tool definitions are removed from the request, preventing the model from seeing the tool schemas. This change enables a workflow where:

1. The model can be aware of available tools (via their definitions in the prompt)
2. But is not expected to use them automatically (since `tool_choice='none'`)

This is useful for:
- Models that need to "plan" about available tools before using them in subsequent requests
- Cases where tool descriptions provide useful context even if the tools aren't used
- Situations where you want the model to reference tool capabilities in its response without actually calling them

## Implementation
- Added a new CLI flag `--expand-tools-even-if-tool-choice-none` (default: False)
- Modified the request validation in `protocol.py` to no longer remove tools when `tool_choice='none'`
- Added the new parameter to `OpenAIServingChat` and passed it through from the API server

okdshin committed:
add option "--expand-tools-even-if-tool-choice-none"

Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin committed:
fix format

Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin committed:
fix format

Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin commented:
Hi @aarnphm, I've fixed the CI issues and updated this PR. The tests are now passing - could you please take another look when you have a chance? Thanks!

aarnphm reviewed (commented):
Do you have an use-case that requires the tool definition even if `tool_choice="none"`?

From a high level POV, having tool definition when I don't want to use tool seems like a waste of context for small/medium size model

okdshin commented:
Currently we are achieving tool_choice="auto" behavior by sending two requests from reverse proxy server to vLLM's API server:

- tool_choice="none" and guided_json={list_of_one_of_signature_of_tools}
- tool_choice="none"

If the first response was an empty json list, we assume the tool use was unnecessary and just use the second response. This is required because we want to use guided decoding (instead of relying on tool_parser) to enforce that the function call strictly aligns with the expected signature.

aarnphm commented:
fyi `'tool_choice="required"'` uses structured outputs here.

okdshin commented:
> I'm not a big fan of the new option,

Honestly, I agree with @russellb . Instead of adding a new option, could we just change the default behavior to be more consistent by always including tool definitions regardless of tool_choice setting?

> fyi 'tool_choice="required"' uses structured outputs here.

Yes. `required` has already supported generally and has almost same behaviour when response is not empty list. `auto` is more complecated.

okdshin commented:
I found the discussion about why tools are dropped when tool_choice="none": https://github.com/vllm-project/vllm/pull/10000#discussion_r1832102569
Dropping tools seems like overkill because it creates inconsistency with other modes (named-function, required, and auto).
The modification here should be sufficient to skip tool parsing when the model hallucinates:
https://github.com/vllm-project/vllm/pull/10000/files#diff-190c665c438d34a7190da9a4d9bc1ed24bed8b13ee1b3f20c6da5c8aa52b0f3bR475-R476 .

okdshin committed:
remove the new option and change the default behaviour

Signed-off-by: okada shintarou <okada@preferred.jp>

russellb reviewed (approved):
I like this change, though I'll hold off on merging to give others more time to express an opinion

aarnphm reviewed (changes requested):
I afraid that this is breaking, given that requests relying on the prior behaviour would start breaking if the set of prompts + user message exceed the max context length of the model (unless long context like YaRN are being used).

Intuitively, `tool_choice="none"` implies that this request won't include any tool definitions. The current behaviour also aligns with OpenAI's logic for `tool_choices="none"` (see [here](https://platform.openai.com/docs/api-reference/responses/create#responses-create-tool_choice))

> `none` means the model will not call any tool and instead generates a message.

This doesn't mean that we have to strictly follow what OpenAI does. Just that if the goal is to be compatible, then I don't think this behaviour makes sense.

aarnphm commented:
opt-out is breaking right? For any production deployment. I think we should make this opt-in, then follow deprecating policy fwiw then break it in 0.10 at least.

@simon-mo do we have plans to release any patches before 0.10?

okdshin committed:
Revert "remove the new option and change the default behaviour"

This reverts commit 377f4ac3d47d67ca05ed2df6054aab6faec02b4d.

Signed-off-by: okada shintarou <okada@preferred.jp>

aarnphm commented:
> Regarding the transition to 0.10 - could you help me understand the process for: What's the recommended way to add deprecation warnings for this kind of behavioral change? Are there any specific documentation or migration guide requirements we should prepare?

https://docs.vllm.ai/en/latest/contributing/deprecation_policy.html is the deprecation policy here. You can search for a few items in the codebase (`enable-reasoning` is one example I can think of)

for 0.10, we are removing v0, so probably a good chance to introduce any breaking change there. See https://github.com/vllm-project/vllm/issues/18571 for the plan

aarnphm reviewed (approved):
stamp from me, but will need to get @russellb's opinion on this if this is reasonable.

russellb reviewed (changes requested):
To follow the deprecation process, we should add some warnings.

1. The help text for the option should clarify that this is a transition option and will be removed in release 0.X.Y.
2. We should detect the case where tools are specified, tool choice is none, and this new option is not in use. We should log a warning in that case.
3. Docs about tool calling should be updated to describe this behavior and how it will change in future releases.

1 and 2 help the vllm admin. 3 is the best I can think of to warn API users right now.

okdshin committed:
update help

Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin committed:
add warn

Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin committed:
fix docs

Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin commented:
@aarnphm @russellb I've addressed the review feedback by adding the deprecation warning https://github.com/vllm-project/vllm/pull/17177/commits/1ffd7f0de6a741eae251ff99dbe02ee852c2ed9c , updating the help text https://github.com/vllm-project/vllm/pull/17177/commits/e3733a97e36e99e2f72e06cef32d89a02b15e8be , and documenting the behavior change https://github.com/vllm-project/vllm/pull/17177/commits/e282ebb0b46ae1b18cec000d3cdfe14ff1a0ad61 . Could you take another look?

okdshin committed:
fix format

Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin committed:
fix format

Signed-off-by: okada shintarou <okada@preferred.jp>

aarnphm commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -172,9 +175,25 @@ async def create_chat_completion(
+                tool_dicts = None
+            elif (request.tool_choice == "none"
+                  and not self.expand_tools_even_if_tool_choice_none):
+                assert request.tools is not None
```
Let's not use assert in performance path here, if this is mostly for types the we can gate it in `TYPE_CHECKING`.

okdshin replied:
@aarnphm Thanks for the review! I understand the performance concern, but I'd like to keep the assert here for a specific reason.

This assert serves as a defensive programming guard rather than just type checking. The logic is:

1. First condition: `if request.tools is None` → `tool_dicts = None`
2. Second condition: `elif (request.tool_choice == "none" and not self.expand_tools_even_if_tool_choice_none)`

The assert ensures that if someone modifies the first condition in the future (e.g., adds another OR condition), we'll catch the logic error immediately with a clear AssertionError, rather than getting a confusing `AttributeError: 'NoneType' object has no attribute '__len__'` when we call `len(request.tools)` below.

While I understand the performance concern, in the context of vLLM's request processing pipeline, this single assertion check is dwarfed by the actual bottlenecks like model inference, GPU operations, and network I/O. The cost of one conditional check per request is negligible compared to the milliseconds/seconds spent on actual LLM processing.

Given that trade-off, I think the defensive programming benefit outweighs the minimal performance cost. What do you think?

aarnphm reviewed (changes requested):
I'm good with this, can  you look at the PR failure? seems like it is relevant to this PR.

okdshin commented on vllm/entrypoints/openai/cli_args.py:
```diff
@@ -223,6 +223,16 @@ def make_arg_parser(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
+        "--expand-tools-even-if-tool-choice-none",
+        action="store_true",
+        default=False,
+        help="[DEPRECATED] Include tool definitions in prompts "
```
```suggestion
        help="Include tool definitions in prompts "
```

okdshin committed:
Apply suggestions from code review

Co-authored-by: Aaron Pham <Aaronpham0103@gmail.com>
Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin committed:
remove assert

Signed-off-by: okada shintarou <okada@preferred.jp>

okdshin commented:
@aarnphm @russellb I've addressed the review feedback. I have no idea why CI failed, but there seems no failure now. I also tested `tests/entrypoints/openai/test_completion_with_prompt_embeds.py` on my local environment and ensured it passed.  Could you take another look?

aarnphm reviewed (approved):
Ok, stamp from me, waiting for @russellb

okdshin commented:
Hi @russellb,

Could you please take a look at the latest changes? The PR has been updated based on your previous suggestions about following the deprecation process. I believe this is now ready for your review.
If anything else needs to be fixed, just let me know - happy to make any additional changes.

russellb reviewed (approved):
apologies for the delay. looks good, thank you!
